agentendpoint/config_task.go

```diff
@@ -0,0 +1,462 @@
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package agentendpoint
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/GoogleCloudPlatform/osconfig/clog"
+
+	agentendpointpb "github.com/GoogleCloudPlatform/osconfig/internal/google.golang.org/genproto/googleapis/cloud/osconfig/agentendpoint/v1alpha1"
+)
+
+const (
+	numExecutionSteps       = 4
+	validationStepIndex     = 0
+	checkStateStepIndex     = 1
+	enforcementStepIndex    = 2
+	postCheckStateStepIndex = 3
+)
+
+type configTask struct {
+	client *Client
+
+	TaskID    string
+	Task      *applyConfigTask
+	StartedAt time.Time `json:",omitempty"`
+
+	// ApplyConfigTaskOutput result
+	results           *agentendpointpb.ApplyConfigTaskOutput_ConfigAssignmentResults
+	postCheckRequired bool
+	assignments       map[string]*assignment
+}
+
+type applyConfigTask struct {
+	*agentendpointpb.ApplyConfigTask
+}
+
+type assignment struct {
+	policies map[string]*policy
+}
+
+type policy struct {
+	resources map[string]*resource
+	hasError  bool
+}
+
+type resource struct {
+	enforcementNeeded bool
+
+	packageResource        *agentendpointpb.PackageResource
+	repositoryResource     *agentendpointpb.RepositoryResource
+	execResource           *agentendpointpb.ExecResource
+	fileResource           *agentendpointpb.FileResource
+	extractArchiveResource *agentendpointpb.ExtractArchiveResource
+	serviceResource        *agentendpointpb.ServiceResource
+}
+
+func (r *resource) unmarshal(ctx context.Context, res *agentendpointpb.ApplyConfigTask_Config_Resource) error {
+	// TODO: implement
+	return nil
+}
+
+func (r *resource) checkState(ctx context.Context) error {
+	// TODO: implement
+	r.enforcementNeeded = false
+	return nil
+}
+
+func (r *resource) runEnforcement(ctx context.Context) error {
+	// TODO: implement
+	return nil
+}
+
+func (c *configTask) reportCompletedState(ctx context.Context, errMsg string, state agentendpointpb.ApplyConfigTaskOutput_State) error {
+	req := &agentendpointpb.ReportTaskCompleteRequest{
+		TaskId:       c.TaskID,
+		TaskType:     agentendpointpb.TaskType_APPLY_CONFIG_TASK,
+		ErrorMessage: errMsg,
+		Output: &agentendpointpb.ReportTaskCompleteRequest_ApplyConfigTaskOutput{
+			ApplyConfigTaskOutput: &agentendpointpb.ApplyConfigTaskOutput{State: state, ConfigAssignmentResults: c.results},
+		},
+	}
+	if err := c.client.reportTaskComplete(ctx, req); err != nil {
+		return fmt.Errorf("error reporting completed state: %v", err)
```

Comment:
Do we need to qualify this error message is for config task?

Reply:
It already will be based on the callers adding error context